Add random 'shadow route' CLTV delta offsets to improve privacy.

tnull opened the pull request:
This PR is  an attempt to address #158. 

In order to generate a plausible CLTV delta offset, a three-hop random walk starting from the payee is conducted and `cltv_expiry_delta` values are aggregated. This offset is then added to all but the final hop of the selected paths (since the last hop knows anyways that it is the recipient). 

Tests are currently failing since they of course do not expect the randomly added offset. I'll fix them when someone had a look at the implementation itself.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1498,6 +1501,55 @@ where L::Target: Logger {
+
+	// Do a three-hop random walk
+	for _random_hop in 0..3 { 
+		if let Some(cur_node) = network_nodes.get(&cur_node_id) {
```
This is great, but I think we should handle the case where the last hop is unknown to us, at least. Consider when you're paying a mobile node through a large LSP, the last-hop will ~never be in the network grpah, but we do want to fudge things starting from the second-to-last hop, pretending we're routing through them. To handle that we'll need to tweak quite a bit here, we may need to calculate such random walks per-path, not just in total. That said, I think we need to *avoid* doing things per-path if the last hop is public - its totally possible the second-to-last public hop will be the same across two paths, which will cause us to tell that node explicitly that we're faking the CLTVs, given they're different on the same hash HTLC.

tnull replied:
Ah, right, I didn't account for MPP and private channels here. Maybe a good way to fix this would be to always apply the offset on a per-path basis, but set the starting point to the predecessor of `payee` when itself is not present in the public graph. The randomness used for the random walk could then come from something like `random_bytes XOR cur_node_id`, which would make the offset the same if all paths share the same starting point and different on a per-path basis otherwise.

TheBlueMatt replied:
Yea, that seems clever, I like that idea.

tnull replied:
I now simply used the first 8 bytes of `cur_node_id` as the `ChaCha20` nonce, which should give us the wanted semi-determinism. See https://github.com/lightningdevkit/rust-lightning/commit/ea0ad9d079ba51e5cf6f780999661a1b7673351b.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1498,6 +1501,55 @@ where L::Target: Logger {
+	let mut random_index: usize;
+
+	// Do a three-hop random walk
+	for _random_hop in 0..3 { 
```
Should we randomize the number here instead of always using 3?

tnull replied:
We definitely could draw it from [0..3] or so, but I'm not entirely sure what would be gained by this. I chose length 3 simply because it roughly reflects current avg. path lengths in the network and therefore should offer an anonymity set corresponding to most of the network.

TheBlueMatt replied:
I worry just always picking 3 reduces our anonymity set - instead of a node knowing the recipient is the last hop, they now know that the recipient is 3 before the last, as calculated using CLTV. Sure, we pick a plausible last-hop, but there may only be one option with the exact CLTV delta we pick. If we pick in the range [1..3] then even if there's only one plausible path with the exact CLTV we use an intermediary still only learns that the recipient is one of 3 nodes.

tnull replied:
Hm, I think since backtracking and loops aren't prevented during the random walk and basically the whole network should be reachable in 3 hops, the anonymity set should not be all that different to begin with. However, I agree that the randomization may least not hurt and is likely useful also for another reason: it may allow us increase/maintain uncertainty while keeping the added 'shadow route' shorter and the CTLV offset lower. This is likely something we want since adding a larger offset may increase the payment failure probability.

TheBlueMatt replied:
> basically the whole network should be reachable in 3 hops

Ah, that's a good point. I guess my concern was that we're picking specific CLTV values based on real nodes, and if we happen to pick three nodes that have something other than super common CLTVs then we're basically announcing who we are again. We could do something much fancier to try to reduce this likelihood, or we could just do 1-3 instead of 3 :)

tnull replied:
Implemented the randomization with https://github.com/lightningdevkit/rust-lightning/commit/ea0ad9d079ba51e5cf6f780999661a1b7673351b. The question on a reasonable upper-limit for the offset remains for me.

TheBlueMatt replied:
Yea, I dunno what the answer is, common node CLTV difference values are 144, right? On the upper end? Given we'd prefer to just use "network-common values" we should just limit to 3*144, no? We could also apply no limit and rely on the `PaymentParameters::max_total_cltv_expiry_delta` setting, which defaults to the total path being limited to a full week.

tnull replied:
I just had another look at ClTV value data I had attached to #1234: the most prevalent value by far is 40 (count: 116133), with 144 only being second (count: 9608). However, as 144 is also our default, I'll probably go with 3*144 for the limit, even though it might technically provide a smaller anonymity set. I think relying on `max_total_cltv_expiry_delta` is not an option here, since this limit is checked and enforced during path finding, i.e., before we add the offset. When we don't have a limit, we might just risk that the payment might get rejected by intermediates during routing itself.

TheBlueMatt replied:
> I think relying on max_total_cltv_expiry_delta is not an option here, since this limit is checked and enforced during path finding, i.e., before we add the offset

Right, but we do need to ensure we never violate `max_total_cltv_expiry_delta` on the full route, including the offset, as that's the guarantee we made to the user. Maybe we should reduce the max we use during pathfinding by 2*144 and then refuse to overrun the cap when we add the offset?

TheBlueMatt replied:
Well if the median is 40, we can also just reduce the cap by 2*40 and then go from there? That way we're at least being relatively conservative.

tnull replied:
Alright, but then we probably also want to use 40 also as the default fallback value for private paths, which would increase the likelihood to stay below the offset limit and the total limit.

tnull replied:
https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/3794bf60245095e06dce60f2babe4e46e6920ecd implements what we discussed so far: 1) subtract `2*40` from the `max_total_cltv_expiry_delta` limit to account for the offset added later. 2) limit the max offset to `3*144`. 3) When we're handling a entirely private path, we add a random offset `x*40, x in [1,...3]`.

Two comments: 1) wouldn't it be 'more conservative' to exclude more failure possibilities from the beginning, i.e., reduce the `max_total_cltv_expiry_delta` value used during path finding by`2*144`? 2) In the case that the CLTV delta  exceeds the total limit after we add the offset, the pathfinding currently just fails with a `LightningError`. Maybe we should at least make this happen less often by applying the random offset and the check not on the final `selected_paths`, but rather before on `drawn_routes`, i.e., swapping Step (9) and Step (10)?

TheBlueMatt replied:
> 1) wouldn't it be 'more conservative' to exclude more failure possibilities from the beginning, i.e., reduce the max_total_cltv_expiry_delta value used during path finding by2*144? 2) In the case that the CLTV delta exceeds the total limit after we add the offset, the pathfinding currently just fails with a LightningError

Ah, I'd been thinking that instead of failing at that point we'd simply cap how much additional CLTV we add. ie if we find a path that is exactly 2*40 less than our `max_total_cltv_expiry_delta` we'd cap our shadow route offset at 80, even if we pick a shadow route higher.

tnull replied:
Ah I see, so the random offset would only be applied when there is still 'room left' beneath the `max_total_cltv_expiry_delta` limit. The issue I see with that (and with enforcing a fixed limit in general) is that there may be some paths for which we only ever can add one or none hop's worth of randomness, which might significantly reduce the anonymity set. In particular in the case where no offset is added at all, the second to last hop on the path could infer that it's successor is the payee, either because the remainder exactly matches the publicly announced `cltv_expiry_delta`, or because it is so small that it's improbable that another hop 'fits in there' so to speak.

TheBlueMatt replied:
Right, there's obviously a tradeoff here, I'll leave it up to you which direction we go (restricting the max during routing more, or making the privacy offset more restricted), but I don't think failing to find a route just because of the privacy offset is an acceptable solution.

tnull replied:
https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/4e9cb21af935e45bee51d4dee8790459a995e38e is a move in this direction: limit the overall applied offset, but don't fail the pathfinding. 
One question that's still open for me, is, whether we should make sure that the offset is a multiple of `40` or `144` when we do indeed limit it, or whether this degree of obfuscation is not worth the effort.

TheBlueMatt reviewed (commented):
Sorry for the delay on this one! A few comments, don't worry about the failing tests for now, I can help out with those if needed.

tnull commented on lightning-invoice/src/utils.rs:
```diff
@@ -129,12 +129,13 @@ where
 pub struct DefaultRouter<G: Deref<Target = NetworkGraph>, L: Deref> where L::Target: Logger {
 	network_graph: G,
 	logger: L,
+	random_bytes: [u8; 32],
```
Thanks for having a look at this! No, very likely we don't want to re-use the randomness. However, I also found it pretty messy to expose the `random_bytes`  to through every part of the higher-up lightning-invoice API to force a user to supply a fresh value every time. Maybe instead of taking `random_bytes: [u8: 32]`, `find_route` could also take a reference to a stateful `prng: ChaCha20`. This would ensure we'd always use different randomness and could also be used to implement a 'real' shuffle in https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L1418.

TheBlueMatt replied:
Alternatively, `DefaultRouter` could store a `ChaCha20` in a mutex and get fresh randomness for each call, no?

tnull replied:
Yes, this of course also works, but we'll then still need to initiate another `ChaCha20` in `get_route` itself. It probably won't hurt though, so I'll go that route.

tnull replied:
Addressed in https://github.com/lightningdevkit/rust-lightning/commit/ea0ad9d079ba51e5cf6f780999661a1b7673351b.

tnull commented on lightning/src/routing/router.rs:
```diff
@@ -1498,6 +1501,55 @@ where L::Target: Logger {
+		if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+			// Randomly choose the next hop
+			prng.process_in_place(&mut random_index_bytes);
+			random_index = usize::from_be_bytes(random_index_bytes).wrapping_rem(cur_node.channels.len());
```
Ah, good point. I think this would currently never be the case since no channels would mean that no path would be found and the function would return in https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L1396. However, its definitely safer to check!

tnull replied:
Ah, good point! If I'm not mistaken `channels.len() == 0` could only ever happen when the last hop is reached through a private channel. I'll hence address this together with your next comment.

tnull replied:
Addressed with https://github.com/lightningdevkit/rust-lightning/commit/ea0ad9d079ba51e5cf6f780999661a1b7673351b.

tnull commented:
Addressed some comments and rebased on https://github.com/lightningdevkit/rust-lightning/commit/a1fedeaec2c551ac4bb3cb2362daef3d81f9e765.

tnull commented on lightning/src/routing/router.rs:
```diff
@@ -1520,14 +1520,13 @@ where L::Target: Logger {
-			assert!(path.len() > 1);
-			NodeId::from_pubkey(&path.iter().rev().next().unwrap().as_ref().unwrap().pubkey)
-		};
+		// Choose the last publicly known node as the starting point for the random walk
```
Using the last publicly known node as the starting point for the random walk is better than the solution before, but still not sufficient: how should we deal with entirely private paths? Better not add any random offset at all in this case?

TheBlueMatt replied:
Hmm, good question, I guess we can always add some fixed value? Like rand(1-3)*144 or something?

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1512,6 +1515,66 @@ where L::Target: Logger {
+		let mut cur_node_id = NodeId::from_pubkey(&cur_node_iter.next().unwrap().as_ref().unwrap().pubkey);
+		assert_eq!(cur_node_id, payee_node_id);
+		while !network_nodes.contains_key(&cur_node_id) {
+			cur_node_id = NodeId::from_pubkey(&cur_node_iter.next().unwrap().as_ref().unwrap().pubkey);
```
Right, this panics if the full route is private, right?

tnull replied:
Yep, this is one reason why I deemed this approach 'insufficient'. Fixed in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/48c7e95ee1a7861681ea93f13aca42f2d835afdf.

tnull commented on lightning/src/util/mod.rs:
```diff
@@ -20,10 +20,10 @@ pub mod errors;
 pub mod ser;
 pub mod message_signing;
 pub mod invoice;
+pub mod chacha20;
```
Ah, I already suspected exposing `ChaCha20` is not the best idea. I agree, iterating the 'initial' random bytes via `sha256` should be sufficient here, especially since `ChaCha20` is still used by `find_route` itself. Addressed this in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/f6793ab2192e57b44801e64900b4e68ae1fd7de1.

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -1512,6 +1516,78 @@ where L::Target: Logger {
+					if let Some(random_index) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len()) {
+						if let Some(random_channel_id) = cur_node.channels.get(random_index) {
+							if let Some(random_channel) = network_channels.get(random_channel_id) {
+								if random_channel.node_one == cur_node_id {
```
One improvement, I think if the randomly selected next hop is actually part of the extended payment path you could pick up another hop to pursue the random walk. What we would like to avoid is the random walk rolling back on the actually drawn out path. Let's say you have A -> B -> C -> D and you sort the random walk D -> C -> B -> A thus a shadow offset partially overlapping with the real cltv deltas. I would say it's a low-probability phenomena but the last publicly known might not have numerous public channels open for routing.

That said, it might not be worthy the effort at that point.

tnull replied:
Thank you very much for having a look at this! Could you elaborate why we wouldn't want the shadow route to overlap the actual path? I see this only as a problem if we assume an on-path adversary that can say for sure that it wasn't part of the actual payment path and hence might be able to reduce the number of candidate routes. However, if we exclude overlapping paths from the beginning, we might reduce the randomness from the getgo, making any adversary's life easier?

tnull replied:
Hum, I think it may not be worth to rule out any overlap, since this could indeed reduce the variety of potential random walks quite a bit. But maybe it would be enough to ensure that we do not immediately loop back on the actual path. And in the case this excludes the only public edge we could fall back to a 'synthetic' random offset, just as in the case for fully private paths.

tnull replied:
I think I'm leaning on the side of leaving this as-is for now and potentially having another look at it in a potential follow-up PR. We might want to have a follow-up anyways, for example in order to introduce some kind of value obfuscation, as noted in #158.

ariard replied:
Yeah really up to you, fine to leave it as a follow-up.

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -1512,6 +1516,78 @@ where L::Target: Logger {
+			shadow_ctlv_expiry_delta_offset = random_walk_length.wrapping_mul(40);
+		}
+
+		// Limit the offset to reduce the payment failure probability
```
I'm not sure if limiting the offset reduce the payment failure, i.e the HTLC not being settled by a preimage delivery. Instead, it does increase the channel closing costs in case of irresponsive counterparty along the path, wasting the liquidity locked along the path. 

Maybe it could say "Limit the offset to reduce the worst-case liquidity timevalue in case of defaulting intermediary hop".

tnull replied:
Yes, adding a large offset would increase the worst-case liquidity cost. For this reason, intermediaries also enforce upper CLTV delta limits during payment forwarding (e.g., LND seems to cap it at 2016 https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/link.go#L49). If I'm not mistaken, we would therefore also risk a payment failure, if we would add an unreasonable large offset. 

But you are right, the first and central reason for capping the offset should be our own liquidity. I'll see to include a more descriptive comment.

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -1512,6 +1516,78 @@ where L::Target: Logger {
+						.checked_add(shadow_ctlv_expiry_delta_offset).unwrap_or(hop.cltv_expiry_delta);
+					let max_total_cltv_expiry_delta = payment_params.max_total_cltv_expiry_delta;
+					if hop.cltv_expiry_delta > max_total_cltv_expiry_delta { 
+						return Err(LightningError{err: "Chosen path exceeds max total CLTV delta limit".to_owned(), action: ErrorAction::IgnoreError});
```
One alternative instead of returning an error would be to compute first the sum of the CLTV deltas along the payment path and ensure if added to the sum of the offsets it's not more than the path upper bound. If so, the overhead can be divided by the number of hops and the result subtract on each offset.

This is motivated by the fact that 432, isn't that much. IIRC Eclair nodes are (were?) using 144 as a default CLTV delta value.

tnull replied:
Yes, this would go hand-in-hand with [Matt's comment](https://github.com/lightningdevkit/rust-lightning/pull/1286#discussion_r806135532) above, i.e., we would enforce a hard limit and essentially distribute the available 'offset budget' equally over all hops. This could be a feasible way to go forward, however, as already stated there I'm generally a bit critical of enforcing a hard limit, since it may give adversaries additional (potentially even certain) information on the payee. Also, if too 'crooked' values would result from splitting the 'offset budget', an adversary may rightfully be able to reduce the candidate set. So we probably should stick with offsets that are multiples of 40 or 144.

Btw. the `432` value is not thought to be the overall CLTV delta limit, which currently defaults to `1008`, but just an upper bound on the added offset.

TheBlueMatt commented:
Looks like this needs rebase, sorry about that.

tnull commented:
> Looks like this needs rebase, sorry about that.

Rebased.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1512,6 +1515,80 @@ where L::Target: Logger {
+
+			// Pick a random path length in [1 .. 3]
+			prng.process_in_place(&mut random_path_bytes);
+			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
```
I believe this will fail to compile on 32-bit platforms. Probably want a u64::from_be_bytes instead (and below) and then convert to usize.

tnull replied:
Ow, that's a good point. However, as Rust even supports `usize`s down to `u16` and squeezing a `u64` into a 32 bit `usize` wouldn't be lossless anyways, I now opted to simply use a `random_bytes` array of size `::core::mem::size_of::<usize>()` in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/ea06afb642fa3fc02aad1c9ad7466da7498cccc7.

TheBlueMatt reviewed (commented):
I'm basically happy with this. There's some compilation errors due to new tests that weren't updated on rebase and the benchmark-only code needs updating, I believe. I'm happy to help resolve any test failures left after that.

TheBlueMatt commented:
To make the router tests pass, maybe it'd be best to split `get_route` into two methods, one that gets the route and one that applies the CLTV deltas so that you don't have to update all the tests.

tnull commented:
> To make the router tests pass, maybe it'd be best to split `get_route` into two methods, one that gets the route and one that applies the CLTV deltas so that you don't have to update all the tests.

Good idea, especially given that `get_route` is ~1000 lines already. I now split adding the offset to its dedicated `add_random_cltv_offset` method and added a test. During this I also stumbled across that currently the CLTV expiry deltas coming from `get_route` are not necessarily monotonically decreasing towards the payee. This should probably be considered a bug, right?

Furthermore, for the time being I left `random_seed_bytes` as a parameter to `get_route` instead of removing it and all the changes to the test etc. I think it could be used to implement the 'real' random shuffle in Step (6). If that is wanted I can add this to this PR or open a follow-up.

tnull commented on lightning/src/routing/router.rs:
```diff
@@ -1520,12 +1529,91 @@ where L::Target: Logger {
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, 
+												   max_path_offset.checked_sub(max_path_offset.wrapping_rem(40)).unwrap_or(max_path_offset));
+
+		// Add 'shadow' CLTV offset to all hops but the final one
```
After thinking about it a bit more, I'm starting to reconsider whether adding the offset to all but the final hop might be the wrong approach? 
As https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L48 describes the `RouteHop`s returned from `get_route` hold the `cltv_expiry_delta` for that specific hop that will be added/aggregated later? Therefore adding the offset for the entire route once to the first hop should suffice here?

TheBlueMatt replied:
Uh, yes, I think you're correct, we should add the CLTV delta only on the last hop (I assume - we want the final recipient to get the extra CLTV, no, not the hop-before-last?)

tnull replied:
Ah, right! I was thinking that adding an offset to the last one would be superfluous since the payee knows anyhow that it's the final hop. However, it's of course correct that the last hop is the one *to* the payee, i.e., the one that the second-to-last node can still observe. Fixed with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/67137c6fb5cf803dcd7031dab6eb0c39bd9fdbca.

TheBlueMatt commented:
Feel free to squash the commits down to one or two, note that we'd ideally like to ensure all code compiles and all tests pass after each commit.

TheBlueMatt commented on lightning-invoice/src/utils.rs:
```diff
@@ -236,8 +239,10 @@ where L::Target: Logger {
+		) -> Result<Route, LightningError> {
+		let mut locked_random_seed_bytes = self.random_seed_bytes.lock().unwrap();
+		*locked_random_seed_bytes = sha256::Hash::hash(&*locked_random_seed_bytes).into_inner();
+		find_route(payer, params, &*self.network_graph, first_hops, &*self.logger, scorer, &locked_random_seed_bytes.clone())
```
As is the lock is held through `find_route` which we maybe don't want. Instead, we'll need an explicit `{}` scope to tell the compiler where to drop the `locked_random_seed_bytes`

jkczyz replied:
Should we consider parameterizing `DefaultRouter` by `K: Deref where K::Target: KeysInterface` and call `get_secure_random_bytes` each time? I suppose the interface is too broad, but I wonder if a narrower `Randomness` interface is warranted. No strong opinion here, but it would avoid the mutex.

tnull replied:
I now went the scoping route with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc. However, I agree that introducing a `Randomness` interface (not only here) would be a nicer abstraction than pushing around arrays with `random_seed_bytes`. Possibly a good follow-up PR?

TheBlueMatt replied:
I suppose we could, yea, agree we'd want to do it in a followup, though. I'm a little hesitant to replace a trivial parameter with a full interface, though, as it just seems like overkill, but we can discuss that later.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -607,19 +608,25 @@ fn compute_fees(amount_msat: u64, channel_fees: RoutingFees) -> Option<u64> {
-	first_hops: Option<&[&ChannelDetails]>, logger: L, scorer: &S
-) -> Result<Route, LightningError>
+	first_hops: Option<&[&ChannelDetails]>, logger: L, scorer: &S, random_seed_bytes: &[u8; 32]
+	) -> Result<Route, LightningError>
```
nit: same here as above.

tnull replied:
Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -607,19 +608,25 @@ fn compute_fees(amount_msat: u64, channel_fees: RoutingFees) -> Option<u64> {
-	)
+		route_params.final_cltv_expiry_delta, logger, scorer, random_seed_bytes
+		) {
+		Ok(mut route) => { 
```
nit: here and in a few other places in this PR you introduce whitespace at EOL. A local `git show` should highlight these for you.

tnull replied:
Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1520,12 +1530,89 @@ where L::Target: Logger {
+			.checked_sub(path.iter().fold(0, |acc,b| b.cltv_expiry_delta.max(acc)))
+			.unwrap_or(shadow_ctlv_expiry_delta_offset);
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, 
+												   max_path_offset.checked_sub(max_path_offset.wrapping_rem(40)).unwrap_or(max_path_offset));
```
nit: please indent the second parameter with just a single tab beyond the previous line. Otherwise its rendered further right than it would be if it were just on the same line :)

tnull replied:
Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

TheBlueMatt commented on lightning/src/routing/scoring.rs:
```diff
@@ -40,8 +41,11 @@
 //!     ..ProbabilisticScoringParameters::default()
 //! };
 //! let scorer = ProbabilisticScorer::new(params, &network_graph);
+//! let random_seed = [42u8; 32];
```
Maybe comment out of the lines here with a `#` at the beginning.

jkczyz replied:
Probably could use a fixed value, too. No need for a `KeysManager`.

tnull replied:
Addressed both with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -4856,22 +5001,44 @@ mod tests {
 		}
 	}
+	#[test]
+	fn adds_cltv_offset() {
```
Could definitely use some further tests, maybe at least one that tests that we limit how much CLTV delta we add if we bump up against the maximum configured.

tnull replied:
I now added additional checks to ensure that we always honor `max_total_cltv_expiry_delta`, i.e., don't add any offset if there is 'no room'.

TheBlueMatt reviewed (commented):
Some style nits and minor test comments.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7427,8 +7431,11 @@ pub mod bench {
+				let random_seed_bytes = keys_manager.get_secure_random_bytes();
 				let route = get_route(&$node_a.get_our_node_id(), &payment_params, &dummy_graph,
-					Some(&usable_channels.iter().map(|r| r).collect::<Vec<_>>()), 10_000, TEST_FINAL_CLTV, &logger_a, &scorer).unwrap();
+					Some(&usable_channels.iter().map(|r| r).collect::<Vec<_>>()), 10_000, TEST_FINAL_CLTV, &logger_a, &scorer,&random_seed_bytes).unwrap();
```
nit: add space before `&random_seed_bytes`

tnull replied:
Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -1081,15 +1081,18 @@ macro_rules! get_route_and_payment_hash {
 			.with_features($crate::ln::features::InvoiceFeatures::known())
 			.with_route_hints($last_hops);
 		let scorer = $crate::util::test_utils::TestScorer::with_penalty(0);
+		let keys_manager = $crate::util::test_utils::TestKeysInterface::new(&[0u8;32], bitcoin::network::constants::Network::Testnet);
```
nit: space after semi-colon in array initialization

tnull replied:
Addressed by https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

jkczyz commented on lightning/src/ln/payment_tests.rs:
```diff
@@ -724,10 +726,12 @@ fn get_ldk_payment_preimage() {
 	let payment_params = PaymentParameters::from_node_id(nodes[1].node.get_our_node_id())
 		.with_features(InvoiceFeatures::known());
 	let scorer = test_utils::TestScorer::with_penalty(0);
+	let keys_manager = test_utils::TestKeysInterface::new(&[0u8;32], Network::Testnet);
```
Likewise.

tnull replied:
Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

jkczyz commented on lightning/src/ln/shutdown_tests.rs:
```diff
@@ -77,6 +78,8 @@ fn updates_shutdown_wait() {
 	let chan_2 = create_announced_chan_between_nodes(&nodes, 1, 2, InitFeatures::known(), InitFeatures::known());
 	let logger = test_utils::TestLogger::new();
 	let scorer = test_utils::TestScorer::with_penalty(0);
+	let keys_manager = test_utils::TestKeysInterface::new(&[0u8;32], Network::Testnet);
```
Likewise.

tnull replied:
Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -607,19 +608,25 @@ fn compute_fees(amount_msat: u64, channel_fees: RoutingFees) -> Option<u64> {
-		route_params.final_cltv_expiry_delta, logger, scorer
-	)
+		route_params.final_cltv_expiry_delta, logger, scorer, random_seed_bytes
+		) {
```
FYI: for indentation, I've been loosely following these recommendations when wrapping:

https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/items.md

tnull replied:
Thanks, I'll have a look!

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -719,8 +727,9 @@ where L::Target: Logger {
 			node_info.features.supports_basic_mpp()
 		} else { false }
 	} else { false };
-	log_trace!(logger, "Searching for a route from payer {} to payee {} {} MPP", our_node_pubkey,
-		payment_params.payee_pubkey, if allow_mpp { "with" } else { "without" });
+	log_trace!(logger, "Searching for a route from payer {} to payee {} {} MPP and {} first hops {}overriding the network graph", our_node_pubkey,
+		payment_params.payee_pubkey, if allow_mpp { "with" } else { "without" },
+		first_hops.map(|hops| hops.len()).unwrap_or(0), if first_hops.is_some() { "" } else { "not " });
```
Looks like an inadvertent change / merge resolution?

tnull replied:
Ow, thank you for pointing this out! Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1512,6 +1521,7 @@ where L::Target: Logger {
 		}
 	}
 
+
```
Remove blank line.

tnull replied:
Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1520,12 +1530,89 @@ where L::Target: Logger {
+				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+					// Randomly choose the next hop
+					prng.process_in_place(&mut random_path_bytes);
+					if let Some(random_index) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len()) {
+						if let Some(random_channel_id) = cur_node.channels.get(random_index) {
+							if let Some(random_channel) = network_channels.get(random_channel_id) {
+								if random_channel.node_one == cur_node_id {
```
This may be a bit easier to follow and remove duplicated code by replacing  each `if let Some(...)` with `and_then` and  possibly both `if ...` with a single `map` using `ChannelInfo::as_directed_to`.

tnull replied:
Thanks, I implemented the first part with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc, which is indeed easier to read. 
However, I'm not sure how the `map`/`as_directed_to` combination would work since the latter provides a `DirectedChannelInfo`, which does however not give direct access to the `cltv_expiry_delta`. So from what I understand right now, I'd have to match on `one_to_two`/`two_to_one` anyhow again to get access to the `ChannelUpdateInfo`?

jkczyz replied:
> However, I'm not sure how the `map`/`as_directed_to` combination would work since the latter provides a `DirectedChannelInfo`, which does however not give direct access to the `cltv_expiry_delta`. So from what I understand right now, I'd have to match on `one_to_two`/`two_to_one` anyhow again to get access to the `ChannelUpdateInfo`?

You should be able to use `DirectedChannelInfo::direction` to get the correct `ChannelUpdateInfo`.

jkczyz replied:
I guess that returns an `Option`, so may have to use it with `and_then` depending on you chain everything.

tnull replied:
Ah, thanks, I overlooked that. I definitely need to look things up in the local docset rather than the docs.rs version.
I hope you're happy with the approach in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/22b4eeb25042813ba905f4d7ea68216624f62aeb.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1520,12 +1530,89 @@ where L::Target: Logger {
+				}
+			}
+		} else {
+			// If the entire path is private, choose a random offset from multiples of 40, i.e., the most used cltv_expiry_delta
```
What is meant by "most used" here?

tnull replied:
Right, "which is the most prevalent cltv_expiry_delta currently used in the network" is probably more descriptive (cf. https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc).

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1520,12 +1530,89 @@ where L::Target: Logger {
+		}
+
+		// Limit the total offset to reduce the worst-case locked liquidity timevalue
+		const MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET: u32 = 432; // 3*144
```
Replace `432` with the calculation instead of using a comment.

tnull replied:
Addressed by https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1520,12 +1530,89 @@ where L::Target: Logger {
+
+		// Limit the offset so we never exceed the max_total_cltv_expiry_delta
+		let max_path_offset = payment_params.max_total_cltv_expiry_delta
+			.checked_sub(path.iter().fold(0, |acc,b| b.cltv_expiry_delta.max(acc)))
```
I think you may be able to use `Iterator::max` here in combination with `Iterator::map`.

tnull replied:
Thanks, using `max`/`map` here is indeed quite a bit simpler.

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -1520,12 +1530,89 @@ where L::Target: Logger {
 	Ok(route)
 }
 
+// Add a random 'shadow route' offset to the CLTV expiry delta values to improve privacy.
```
nit: "Only the *final* hop is offset".

ariard replied:
Also nice if you can add a quick note explaining which privacy is improved here.

E.g "to prevent intermediate node to observe the payment destination". Or refer to the BOLT 7 recommendation.

tnull replied:
Thanks, I now tried to include a more descriptive comment regarding the issue!

ariard reviewed (commented):
Good with the new approach of only offsetting the final hop.
(Thought the old one had the benefit to also mask the sender topology as the intermediate node couldn't deduce from their CLTV deltas the distance from sender, I believe)

tnull commented on fuzz/src/router.rs:
```diff
@@ -250,6 +251,7 @@ pub fn do_test<Out: test_logger::Output>(data: &[u8], out: Out) {
 					}
 				}
 				let scorer = FixedPenaltyScorer::with_penalty(0);
+				let random_seed_bytes: [u8; 32] = get_slice!(32).try_into().unwrap();
```
Ah, I didn't think about that but it definitely makes sense. Should be addressed by https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

tnull commented on fuzz/src/full_stack.rs:
```diff
@@ -456,7 +456,8 @@ pub fn do_test(data: &[u8], logger: &Arc<dyn Logger>) {
 					final_cltv_expiry_delta: 42,
 				};
-				let route = match find_route(&our_id, &params, &network_graph, None, Arc::clone(&logger), &scorer) {
+				let random_seed_bytes: [u8; 32] = get_slice!(32).try_into().unwrap();
```
Should be addressed by https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

tnull commented on lightning-background-processor/src/lib.rs:
```diff
@@ -667,13 +667,15 @@ mod tests {
 
 	#[test]
 	fn test_invoice_payer() {
+		let keys_manager = test_utils::TestKeysInterface::new(&[0u8;32], Network::Testnet);
```
Fixed with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc.

tnull commented:
Thanks for all the great feedback, I'll take it into account going forward!

tnull commented:
Hm, after rebasing once more, the tests some of the checks seem to fail due to a [crates.io](https://crates.io) downtime..

TheBlueMatt commented:
> some of the checks seem to fail due to a [crates.io](https://crates.io/) downtime..

Kicked CI so hopefully it re-runs and passes.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1521,12 +1529,91 @@ where L::Target: Logger {
+
+		// Limit the offset so we never exceed the max_total_cltv_expiry_delta
+		let max_path_offset = payment_params.max_total_cltv_expiry_delta
+			.checked_sub(path.iter().map(|h| h.cltv_expiry_delta).max().unwrap_or(0))
```
Shouldn't we sum the `cltv_expiry_delta`s in each hop instead of finding the maxiumum one?

tnull replied:
Thanks for bringing this up. The thought here was that the `max` is even more restrictive of the offset. However, through this I realized I still also used the `max` instead of the `sum` for calculating the `max_path_offset`, which was a remnant from when the offset used to be added to all hops along the route, not only the final one. Fixed now with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/22b4eeb25042813ba905f4d7ea68216624f62aeb.

tnull commented on lightning/src/routing/router.rs:
```diff
@@ -1521,12 +1529,91 @@ where L::Target: Logger {
+			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+
+			// Init PRNG with path nonce
+			let mut path_nonce = [0u8; 8];
```
Great, I now use generous 12 bytes in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/22b4eeb25042813ba905f4d7ea68216624f62aeb! ;)

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1537,12 +1545,85 @@ where L::Target: Logger {
+					if let Some(random_channel) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len())
+						.and_then(|index| cur_node.channels.get(index))
+						.and_then(|id| network_channels.get(id)) {
+							random_channel.as_directed_to(&cur_node_id).map(|(dir_info, next_id)| {
```
Isn't this the wrong direction? We want the channel directional info from the current node to some other node. It'd be nice to have a test which demonstrated if this is right.

tnull replied:
Now added a plausibility test with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/a5b8780b91b890457781411befc54ecd24e19267.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1537,12 +1545,85 @@ where L::Target: Logger {
+		// Limit the offset so we never exceed the max_total_cltv_expiry_delta
+		let max_path_offset = payment_params.max_total_cltv_expiry_delta
+			.checked_sub(path.iter().map(|h| h.cltv_expiry_delta).sum())
+			.unwrap_or(shadow_ctlv_expiry_delta_offset);
```
Shouldn't this be `unwrap_or(0)`? If our path expiry is more than the max already we probably shouldn't add an offset at all? Also, were we planning on limiting the total CLTV during the `get_route` run to give us a little headroom here?

tnull replied:
Well my thinking here was that since we limit the route overall with `max_total_cltv_expiry_delta`, this subtraction should never fail. If however something unexpected would occur and it would fail, we might be better off to default to provide some privacy (capped at `3*144`), instead of none. But it can of course also default to `0`, I have no strong preference here.

As for the considering the offset in `get_route`: [after looking at the data](https://github.com/lightningdevkit/rust-lightning/pull/1286#discussion_r809896779) I found that there were basically no cases in which there was not ample room to add the offset and hence thought it would be cleaner not to interfere with pathfinding and keep the two methods entirely separate. However, if we still want to take some of the offset into account, I can of course go back to, for example, subtracting `2*40` or `2*144` from the limit during path finding.

TheBlueMatt replied:
> Well my thinking here was that since we limit the route overall with max_total_cltv_expiry_delta, this subtraction should never fail.

Right, sorry, indeed.

>  If however something unexpected were to happen and it would fail, we might be better off to default to provide some privacy (capped at 3*144), instead of none.

:man_shrugging: I guess we shouldn't ever violate the user-provided value ever?

> As for the considering the offset in get_route: https://github.com/lightningdevkit/rust-lightning/pull/1286#discussion_r809896779 I found that there were basically no cases in which there was not ample room to add the offset and hence thought it would be cleaner not to interfere with pathfinding and keep the two methods entirely separate.

Hmm, I feel like that's a better reason to interfere than to not interfere.

I'm still a bit confused by this hunk here - why do we do the subtraction at `max_path_offset.checked_sub(max_path_offset.wrapping_rem(40))`?

tnull replied:
> 🤷‍♂️ I guess we shouldn't ever violate the user-provided value ever?
> ...
> Hmm, I feel like that's a better reason to interfere than to not interfere.
Alright, both addressed by https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/f7be9daded106710d969ddfcc6c3a949a05b906c.

> I'm still a bit confused by this hunk here - why do we do the subtraction at `max_path_offset.checked_sub(max_path_offset.wrapping_rem(40))`?
The difference of the path total to `max_total_cltv_expiry_delta` may be a pretty uncommon value, which could result in an implausible shadow route offset. The thinking here therefore was to choose the path offset to be the closest multiple of 40 to increase plausibility. I now had another look at this part and added a comment to make this more transparent in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/f7be9daded106710d969ddfcc6c3a949a05b906c.

jkczyz commented on lightning/src/routing/network_graph.rs:
```diff
@@ -674,6 +674,21 @@ impl ChannelInfo {
+
+	/// Returns a [`DirectedChannelInfo`] for the channel coming from the given `source` to a
+	/// returned `target`, or `None` if `source` is not one of the channel's counterparties.
+	pub fn as_coming_from(&self, source: &NodeId) -> Option<(DirectedChannelInfo, &NodeId)> {
```
Would prefer to name this `as_directed_from` for consistency.

tnull replied:
Sure, addressed with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/d8fe725ae98226dad4aa3c37d7d5650ccdd13ad9.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -222,12 +223,14 @@ where
-	/// Creates a new router using the given [`NetworkGraph`] and  [`Logger`].
-	pub fn new(network_graph: G, logger: L) -> Self {
-		Self { network_graph, logger }
+	/// Creates a new router using the given [`NetworkGraph`], a [`Logger`], and a randomness source `random_seed_bytes`.
```
Wrap comments at 100 characters.

tnull replied:
Done with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/7c5ee057822dcb690c542e6aa0ea87fb4882a266.

jkczyz commented on lightning-invoice/src/utils.rs:
```diff
@@ -237,7 +240,13 @@ where L::Target: Logger {
 		first_hops: Option<&[&ChannelDetails]>, scorer: &S
 	) -> Result<Route, LightningError> {
-		find_route(payer, params, &*self.network_graph, first_hops, &*self.logger, scorer)
+		let random_seed_bytes: [u8; 32];
+		{
+			let mut locked_random_seed_bytes = self.random_seed_bytes.lock().unwrap();
+			*locked_random_seed_bytes = sha256::Hash::hash(&*locked_random_seed_bytes).into_inner();
+			random_seed_bytes = *locked_random_seed_bytes;
+		}
```
Note this could also be written as:

```rust
let random_seed_bytes = {
    // ...
   *locked_random_seed_bytes
};
```

tnull replied:
Indeed much cleaner. Done with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/7c5ee057822dcb690c542e6aa0ea87fb4882a266.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -853,7 +861,10 @@ where L::Target: Logger {
+					// We subtract 2*40 here in order to account for some of the privacy-enhancing
+					// random CLTV expiry delta offset we add on top later.
+					let max_total_cltv_expiry_delta = payment_params.max_total_cltv_expiry_delta
+						.checked_sub(2*40).unwrap_or(payment_params.max_total_cltv_expiry_delta);
```
Is the significance of `2*40` mentioned anywhere? Could you add a constant for `40` since it's used elsewhere?

tnull replied:
No, it's a rough estimate of the random walk's expected value, since `40` is the most commonly used value in the network today and the random walk has a length in [1..3]. I now added an explanation and a `MEDIAN_HOP_CLTV_EXPIRY_DELTA` constant with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/4446516b4128ecc4f52d7ab991f35208fd635df6.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1537,12 +1548,87 @@ where L::Target: Logger {
+// graph. In order to improve privacy, this method obfuscates the CLTV expiry deltas along the
+// payment path by adding a randomized 'shadow route' offset to the final hop.
+fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters, network: &NetworkGraph, random_seed_bytes: &[u8; 32]) {
+	let network_graph = network.read_only();
+	let network_channels = network_graph.channels();
+	let network_nodes = network_graph.nodes();
```
I'm a little concerned that we lock these twice. Once here and once prior to this in `get_route`. It's possible the `NetworkGraph` could change in the interim. I think we want to call `add_random_cltv_offset` from within `get_route` instead such that a single read-only lock is used.

tnull replied:
I understand that locking twice is generally suboptimal. However, calling `add_random_cltv_offset` from `get_route` is probably not what we want, since we split them off to avoid having to make all dependent tests account for some random variety in the routes returned. This would not only entail a lot of changes, but also would require the testcases to be less strict. Maybe a middleground would be to lock the graph once in `find_route` and let `get_route` and `add_random_cltv_offset` take the returned `ReadOnlyNetworkGraph`?

jkczyz replied:
> Maybe a middleground would be to lock the graph once in `find_route` and let `get_route` and `add_random_cltv_offset` take the returned `ReadOnlyNetworkGraph`?

Yeah, I'm good with that. At one point `get_route` was the only public interface but we refactored it behind `find_route`.

tnull replied:
Great! Went this route with https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/521ae09a03718c17cdafe52644213ce5c8fac134.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1537,12 +1548,87 @@ where L::Target: Logger {
+						.and_then(|index| cur_node.channels.get(index))
+						.and_then(|id| network_channels.get(id)) {
+							random_channel.as_coming_from(&cur_node_id).map(|(dir_info, next_id)| {
+								dir_info.direction().map(|channel_update_info|
```
We use `DirectedChannelInfoWithUpdate` with `PublicHop`, so I believe this should always be `Some`. Using `map` is safer if the code changes, but `unwrap` should suffice. Not sure if it really matters much, but just wanted to note this.

tnull replied:
Ah, thanks, good to know!

tnull commented:
Rebased and squashed once more.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -623,19 +626,26 @@ fn compute_fees(amount_msat: u64, channel_fees: RoutingFees) -> Option<u64> {
-	logger: L, scorer: &S
+	our_node_pubkey: &PublicKey, payment_params: &PaymentParameters, network_graph: &ReadOnlyNetworkGraph,
+	first_hops: Option<&[&ChannelDetails]>, final_value_msat: u64, final_cltv_expiry_delta: u32,
+	logger: L, scorer: &S, random_seed_bytes: &[u8; 32]
```
`random_seed_bytes` is not needed here since it is only used in `add_random_cltv_offset`.

tnull replied:
That is correct, but the idea was to leave it in `get_route` for now and use it to implement the ['real random shuffle' TODO](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/routing/router.rs#L1439) in a follow-up PR. I now prefixed it with an underscore to disable the warning, but let me know if you'd rather keep it out of `get_route` altogether in this PR.

jurvis replied:
hey @tnull is it okay if I take a stab at implementing the real random shuffle? I believe it may be related to this issue https://github.com/lightningdevkit/rust-lightning/issues/869

tnull replied:
> hey @tnull is it okay if I take a stab at implementing the real random shuffle? I believe it may be related to this issue #869

Ah, thank you for the interest! This issue indeed seems related. However, I’m already mostly done with a first draft and plan to open a PR in the next few days.

jkczyz commented on fuzz/src/router.rs:
```diff
@@ -32,6 +32,7 @@ use utils::test_logger;
 use std::collections::HashSet;
 use std::sync::Arc;
 use std::sync::atomic::{AtomicUsize, Ordering};
+use std::convert::TryInto;
```
Not used.

tnull replied:
Removed in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/f5b1b46c43d740cf531f9f4ae669303043680cf0.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -1535,17 +1550,93 @@ where L::Target: Logger {
+		let mut max_path_offset = payment_params.max_total_cltv_expiry_delta
+			.checked_sub(path.iter().map(|h| h.cltv_expiry_delta).sum())
+			.unwrap_or(0);
+		max_path_offset = max_path_offset
+			.wrapping_sub(max_path_offset.wrapping_rem(MEDIAN_HOP_CLTV_EXPIRY_DELTA))
+			.max(max_path_offset.wrapping_rem(MEDIAN_HOP_CLTV_EXPIRY_DELTA));
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, max_path_offset);
```
If the subtraction wraps, `shadow_ctlv_expiry_delta_offset` would not likely change, IIUC. Could it then possibly exceed `payment_params.max_total_cltv_expiry_delta` for small values?

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -4956,6 +5109,104 @@ mod tests {
+		let mut route_default = route.clone();
+		add_random_cltv_offset(&mut route_default, &payment_params, &network_graph.read_only(), &random_seed_bytes);
+		let cltv_expiry_deltas_default = route_default.paths.iter().flat_map(|p| p.iter().map(|h| h.cltv_expiry_delta).collect::<Vec<u32>>()).collect::<Vec<u32>>();
+		assert_eq!(cltv_expiry_deltas_before[.. cltv_expiry_deltas_default.len()-1], cltv_expiry_deltas_default[.. cltv_expiry_deltas_default.len()-1]);
+		assert!(cltv_expiry_deltas_default.last() > cltv_expiry_deltas_before.last());
+		assert!(cltv_expiry_deltas_default.last().unwrap() <= &DEFAULT_MAX_TOTAL_CLTV_EXPIRY_DELTA);
```
Could simplify this a big by using `split_last`.

tnull replied:
Yep, simplified this in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/0b777cbd4ac95d1578264dab82826d725d2cd3d3.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -4956,6 +5109,104 @@ mod tests {
+		let keys_manager = test_utils::TestKeysInterface::new(&[0u8; 32], Network::Testnet);
+		let random_seed_bytes = keys_manager.get_secure_random_bytes();
+		let route = get_route(&our_id, &payment_params, &network_graph.read_only(), None, 100, 42, Arc::clone(&logger), &scorer, &random_seed_bytes).unwrap();
+		let cltv_expiry_deltas_before = route.paths.iter().flat_map(|p| p.iter().map(|h| h.cltv_expiry_delta).collect::<Vec<u32>>()).collect::<Vec<u32>>();
```
Will there only be on path in the route? If so, could we assert that and take `route.paths[0]` and avoid some of this code?

tnull replied:
Alright, simplified this in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/0b777cbd4ac95d1578264dab82826d725d2cd3d3.

tnull commented on lightning/src/routing/network_graph.rs:
```diff
@@ -674,6 +674,21 @@ impl ChannelInfo {
 		Some((DirectedChannelInfo { channel: self, direction }, source))
 	}
+
+	/// Returns a [`DirectedChannelInfo`] for the channel coming from the given `source` to a
```
Addressed in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/0b777cbd4ac95d1578264dab82826d725d2cd3d3..

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -1535,17 +1550,93 @@ where L::Target: Logger {
+			.checked_sub(path.iter().map(|h| h.cltv_expiry_delta).sum())
+			.unwrap_or(0);
+		max_path_offset = max_path_offset
+			.saturating_sub(max_path_offset.wrapping_rem(MEDIAN_HOP_CLTV_EXPIRY_DELTA))
```
A lot of the code here and above would be substantially easier to read without the functional-style math. eg the docs for `wrapping_rem` even explicitly states that its identical to `%`, and above there's a good chunk of math that trivially cannot overflow.

tnull replied:
Hm, sorry for that, I got into the habit of using the 'functional style' consistently throughout, especially when some closely connected parts of the code require it. I should reconsider that since I agree that it's sometimes not as easy to read. I now further simplified the parts in question with https://github.com/lightningdevkit/rust-lightning/commit/e2efb0c4f01db56b4fd5d0b8898d820b55747ecf.

tnull commented:
Rebased.
